packages/server/src/api/controllers/row/index.ts

```diff
@@ -98,12 +100,26 @@ export async function find(ctx: any) {
   })
 }
 
-export async function destroy(ctx: any) {
+export async function destroy(ctx: UserCtx<DeleteRowRequest>) {
   const appId = ctx.appId
-  const inputs = ctx.request.body
+  const inputs: DeleteRowRequest = ctx.request.body
+
   const tableId = utils.getTableId(ctx)
   let response, row
-  if (inputs.rows) {
+
+  if ("rows" in inputs && Array.isArray(inputs?.rows)) {
```

Comment:
I think something that can help the readability here are some type guards

```ts
function isDeleteRows(input: any): input is DeleteRows {
  return input.rows !== undefined && Array.isArray(input.rows);
}

function isDeleteRow(input: any): input is DeleteRow {
  return input._id !== undefined;
}

function isStringRow(row: any): row is string {
  return typeof row === 'string';
}
```

We can then use like
```ts
// multiple delete request
if (isDeleteRows(inputs)) {
  ...
  let processedRow: Row = isStringRow(row) ? 
      { _id: row } : // row id request
      row // row object request
  ...
} 
// singular delete request
else if (isDeleteRow(inputs) {
  ...
}
```

I would advise staying away from things like `"rows" in inputs` as it completely ignores any typed fields that live in the target object and would be missed by refactors. 

The

Reply:
The controller function is getting pretty huge at this point. In an effort to tame this it would help to break out some helper functions as well like:

```ts
function deleteRows(body: DeleteRows) { }
function deleteRow(body: DeleteRow) {}
```

And just life the contents of each section of the `if/else` into them respectively.